documentation/matlab-api.md

```diff
@@ -0,0 +1,381 @@
+end
+
+(from disp2sqw)
+```
+
+#### Pros
+- Completely flexible handling of arguments specific to function
+
+#### Cons
+- Unreadable
+- Unmaintainable
+- Unintuitive
+- Breaks easily
+
+### Comparison
+
+| Type of Arg | `inputParser` | `parse_arguments` | Raw |
+|:-----------:|:-------------:|:-----------------:|:---:|
+| Positional  |      Y        |        Y          |  Y  |
+| Required    |      Y        |        Y          |  Y  |
+| Optional    |      Y        |        Y          |  Y  |
+| Key-value   |      Y        |        Y          |  Y  |
+| Flag        |      N        |        Y          |  Y  |
+| Default val |      Y        |        Y          |  Y  |
+| Validate value  |  Y        |        N          |  Y  |
+| Validate type |    Y        |        N          |  Y  |
+| Abbreviation |     N        |        Y          |  Y  |
+
+## Options
+
+### Dash flags `-(no)flagname`
+
+Allow flags with dashes in MATLAB
+
+#### MATLAB
+
+Currently, using argument parser is flexible in the way it handles flags and allows the same flag (or its negation) to be specified in multiple ways.
+```
+function_name(a, b, c, 'flagname')
+function_name(a, b, c, '-flagname')
+function_name(a, b, c, '-noflagname')
+function_name(a, b, c, '-f')
+function_name(a, b, c, '-nof')
+```
+This flexibility allows less typing, but at the same time may obscure the intended operation or even lead to the wrong flag being used.
+
+
+#### Python
+
+```
+def function_name(a, b, c, flagname = true) #default
+
+def function_name(a, b, *args, **kwargs):
+    if any(arg.startswith('-') for arg in args): # Handle flags
+        pass
+    # args = [value-of-c, ...]
+    # kwargs = {filename: true, ...}
+```
+As Python does not allow args starting with `-`, this will require some way of handling the flag.
+```
+def function_name(a, -flag) # ERROR
+```
+
+This can either be by passing flags through as a string pre-parsing the input before the call to MATLAB to be handled by `*args`
+```
+def function_name(a, b, *args, **kwargs)
+    if any(arg.startswith('-') for arg in args if isinstance(arg, string)): # Handle flags
+       pass
+```
+Or passed directly through as one of args.
+```
+def function_name(a, b, *args, **kwargs):
+    call_MATLAB(a, b, *args, **kwargs)
+```
+
+Which would be called as:
+```
+function_name(a, b, '-flag')
+```
+
+Or by passing in through dict form to be handled by `**kwargs` (nobody would do this).
+```
+function_name(a, b, **{'-flag':True})
+```
+
+This will depend on the ultimate interface `call_MATLAB` has.
+
+#### Pros and Cons
+
+##### Pros
+
+- Less typing for the average user?
+```
+function_name(a, '-flag')
+function_name(a, 'flag', 1)
+function_name(a, '-noflag')
+function_name(a, 'flag', 0)
+```
+- Python can just pass args directly into the MATLAB call in principle
+
+##### Cons
+
+- Raises questions about the separation between arguments and flags are additional arguments to be labelled as flags?
+- Not Pythonic way of doing things
```

Comment:
I think we should be spelling out exactly what is "not Pythonic" in any given instance

Reply:
I have tried to address the Pythonic way of doing things in the appendix

Reply:
In Python I would prefer to use "keyword=True/False" instead of flags, instead of using a string which starts with a minus.

If we're writing (perhaps automatically generated) wrappers for each Matlab function, this could be handled without changing the Matlab code by parsing the Matlab arguments. If we're going to just use generic wrappers this would require changing the Matlab code.

For example, the `-nopix` flag of `cut_sqw` can be specified in Python using the `pix` keyword which is `True` by default. E.g. `cut_sqw(..., '-nopix')` in Matlab would be `cut_sqw(..., pix=False)` in Python.